tunnel-server: server connection closed

I have setup a k8s cluster with the master and a working node on separate networks. I referenced [this tutorial](https://github.com/alibaba/openyurt/blob/master/docs/tutorial/yurt-tunnel.md#5-setup-the-yurt-tunnel-manually) to setup the tunnel sever and agent, but I can't access the pod on edge node through yurt-tunnel. The logs from the tunnel-server:

```
$ kubectl logs yurt-tunnel-server-74cfdd4bc7-7rrmr -n kube-system
I1110 12:53:57.737387       1 cmd.go:143] server will accept yurttunnel-agent requests at: 192.168.1.101:10262, server will accept master https requests at: 192.168.1.101:10263server will accept master http request at: 192.168.1.101:10264
W1110 12:53:57.737429       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1110 12:53:57.968315       1 iptables.go:474] clear conntrack entries for ports ["10250" "10255"] and nodes ["192.168.1.101" "192.168.122.55" "127.0.0.1"]
E1110 12:53:57.992841       1 iptables.go:491] clear conntrack for 192.168.1.101:10250 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1110 12:53:58.011089       1 iptables.go:491] clear conntrack for 192.168.122.55:10250 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1110 12:53:58.025873       1 iptables.go:491] clear conntrack for 127.0.0.1:10250 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1110 12:53:58.035197       1 iptables.go:491] clear conntrack for 192.168.1.101:10255 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1110 12:53:58.042357       1 iptables.go:491] clear conntrack for 192.168.122.55:10255 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1110 12:53:58.048433       1 iptables.go:491] clear conntrack for 127.0.0.1:10255 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
I1110 12:54:03.073595       1 csrapprover.go:52] starting the crsapprover
I1110 12:54:03.209064       1 csrapprover.go:174] successfully approve yurttunnel csr(csr-sqfnw)
I1110 12:54:08.070368       1 anpserver.go:101] start handling request from interceptor
I1110 12:54:08.070787       1 anpserver.go:137] start handling https request from master at 192.168.1.101:10263
I1110 12:54:08.070872       1 anpserver.go:151] start handling http request from master at 192.168.1.101:10264
I1110 12:54:08.071365       1 anpserver.go:189] start handling connection from agents
I1110 12:54:09.087254       1 server.go:418] Connect request from agent ubuntu-standard-pc-i440fx-piix-1996
I1110 12:54:09.087319       1 backend_manager.go:99] register Backend &{0xc000158480} for agentID ubuntu-standard-pc-i440fx-piix-1996
W1110 12:54:24.273510       1 server.go:451] stream read error: rpc error: code = Canceled desc = context canceled
I1110 12:54:24.273532       1 backend_manager.go:119] remove Backend &{0xc000158480} for agentID ubuntu-standard-pc-i440fx-piix-1996
I1110 12:54:24.273562       1 server.go:531] <<< Close backend &{0xc000158480} of agent ubuntu-standard-pc-i440fx-piix-1996
I1110 12:54:37.682857       1 csrapprover.go:174] successfully approve yurttunnel csr(csr-6lcjl)
I1110 12:54:42.969063       1 server.go:418] Connect request from agent ubuntu-standard-pc-i440fx-piix-1996
I1110 12:54:42.969111       1 backend_manager.go:99] register Backend &{0xc000158180} for agentID ubuntu-standard-pc-i440fx-piix-1996
```

Login to the edge node, the tunnel-agent container log indicates a "connection closed" error. Any idea how to solve this issue? Thanks.

```
I1110 12:54:37.583915       1 cmd.go:106] neither --kube-config nor --apiserver-addr is set, will use /etc/kubernetes/kubelet.conf as the kubeconfig
I1110 12:54:37.583964       1 cmd.go:110] create the clientset based on the kubeconfig(/etc/kubernetes/kubelet.conf).
I1110 12:54:37.647689       1 cmd.go:135] yurttunnel-server address: 192.168.1.101:31302
I1110 12:54:37.647990       1 anpagent.go:54] start serving grpc request redirected from yurttunel-server: 192.168.1.101:31302
E1110 12:54:37.657318       1 clientset.go:155] rpc error: code = Unavailable desc = connection closed
I1110 12:54:42.970218       1 stream.go:255] Connect to server 3656d4f5-746f-411f-b0eb-c1c69b1ff2c1
I1110 12:54:42.970241       1 clientset.go:184] sync added client connecting to proxy server 3656d4f5-746f-411f-b0eb-c1c69b1ff2c1
I1110 12:54:42.970266       1 client.go:122] Start serving for serverID 3656d4f5-746f-411f-b0eb-c1c69b1ff2c1
```

> > it looks like the yurt-server has been restarted 14 times, is it normal?
> 
> It is unnormal, Could you do `kubectl logs yurt-tunnel-server-58f9d8cd54-4pg2j --previous` and upload the log?

@charleszheng44, in addition to the above `nat` rules, thanks for the tip to get to the bottom why the yurt-server has restarted periodically, it looks like the server was crashed due to a nil pointer dereference... do you want to make a wild guess what's going on? Thanks.

```
$ kubectl logs yurt-tunnel-server-58f9d8cd54-5kk84 -n kube-system --previous
I1116 09:45:46.292104       1 cmd.go:143] server will accept yurttunnel-agent requests at: 192.168.1.101:10262, server will accept master https requests at: 192.168.1.101:10263server will accept master http request at: 192.168.1.101:10264
W1116 09:45:46.292172       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1116 09:45:46.537568       1 iptables.go:474] clear conntrack entries for ports ["10250" "10255"] and nodes ["192.168.122.55" "192.168.1.101" "127.0.0.1"]
E1116 09:45:46.545064       1 iptables.go:491] clear conntrack for 192.168.122.55:10250 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
I1116 09:45:46.552436       1 iptables.go:495] clear conntrack for 192.168.1.101:10250 successfully: "conntrack v1.4.4 (conntrack-tools): 1 flow entries have been deleted.\ntcp      6 86382 ESTABLISHED src=192.168.1.101 dst=192.168.1.101 sport=59084 dport=10250 src=192.168.1.101 dst=192.168.1.101 sport=10250 dport=59084 [ASSURED] mark=0 use=1\n"
E1116 09:45:46.559758       1 iptables.go:491] clear conntrack for 127.0.0.1:10250 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1116 09:45:46.567230       1 iptables.go:491] clear conntrack for 192.168.122.55:10255 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1116 09:45:46.573038       1 iptables.go:491] clear conntrack for 192.168.1.101:10255 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
E1116 09:45:46.577821       1 iptables.go:491] clear conntrack for 127.0.0.1:10255 failed: "conntrack v1.4.4 (conntrack-tools): 0 flow entries have been deleted.\n", error message: exit status 1
I1116 09:45:51.594407       1 csrapprover.go:52] starting the crsapprover
I1116 09:45:51.703369       1 csrapprover.go:174] successfully approve yurttunnel csr(csr-txtdp)
I1116 09:45:56.595282       1 anpserver.go:101] start handling request from interceptor
I1116 09:45:56.595617       1 anpserver.go:151] start handling http request from master at 192.168.1.101:10264
I1116 09:45:56.595811       1 anpserver.go:189] start handling connection from agents
I1116 09:45:56.595810       1 anpserver.go:137] start handling https request from master at 192.168.1.101:10263
I1116 09:45:58.129948       1 server.go:418] Connect request from agent ubuntu-standard-pc-i440fx-piix-1996
I1116 09:45:58.130011       1 backend_manager.go:99] register Backend &{0xc000138480} for agentID ubuntu-standard-pc-i440fx-piix-1996
E1116 11:45:11.902269       1 csrapprover.go:86] certificatesigningrequest.certificates.k8s.io "csr-txtdp" not found
E1116 11:45:11.902587       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 31 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x146d6e0, 0x22fd960)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/runtime/runtime.go:48 +0x82
panic(0x146d6e0, 0x22fd960)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/api/certificates/v1beta1.(*CertificateSigningRequest).GetNamespace(0x0, 0x0, 0x7fad04412b88)
	<autogenerated>:1 +0x5
k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1617e20, 0x0, 0xc000156320, 0x1, 0x2, 0x22fd7c0)
	/go/pkg/mod/k8s.io/client-go@v0.16.9/tools/cache/store.go:84 +0x10e
github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.enqueueObj(0x18a3b40, 0xc00002abe0, 0x1617e20, 0x0)
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:99 +0x39
github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.(*YurttunnelCSRApprover).processNextItem(0xc000436600, 0x0)
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:87 +0x22a
github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.(*YurttunnelCSRApprover).runWorker(0xc000436600)
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:66 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000322040)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000322040, 0x3b9aca00, 0x0, 0x45a401, 0xc00009a600)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000322040, 0x3b9aca00, 0xc00009a600)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/wait/wait.go:88 +0x4d
created by github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.(*YurttunnelCSRApprover).Run
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:59 +0x24d
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xf81dc5]

goroutine 31 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/runtime/runtime.go:55 +0x105
panic(0x146d6e0, 0x22fd960)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/api/certificates/v1beta1.(*CertificateSigningRequest).GetNamespace(0x0, 0x0, 0x7fad04412b88)
	<autogenerated>:1 +0x5
k8s.io/client-go/tools/cache.MetaNamespaceKeyFunc(0x1617e20, 0x0, 0xc000156320, 0x1, 0x2, 0x22fd7c0)
	/go/pkg/mod/k8s.io/client-go@v0.16.9/tools/cache/store.go:84 +0x10e
github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.enqueueObj(0x18a3b40, 0xc00002abe0, 0x1617e20, 0x0)
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:99 +0x39
github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.(*YurttunnelCSRApprover).processNextItem(0xc000436600, 0x0)
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:87 +0x22a
github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.(*YurttunnelCSRApprover).runWorker(0xc000436600)
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:66 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000322040)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000322040, 0x3b9aca00, 0x0, 0x45a401, 0xc00009a600)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000322040, 0x3b9aca00, 0xc00009a600)
	/go/pkg/mod/k8s.io/apimachinery@v0.16.10-beta.0/pkg/util/wait/wait.go:88 +0x4d
created by github.com/alibaba/openyurt/pkg/yurttunnel/pki/certmanager.(*YurttunnelCSRApprover).Run
	/opt/src/pkg/yurttunnel/pki/certmanager/csrapprover.go:59 +0x24d
```

--- a/pkg/yurttunnel/pki/certmanager/csrapprover.go
+++ b/pkg/yurttunnel/pki/certmanager/csrapprover.go
@@ -22,9 +22,6 @@ import (
 	"fmt"
 	"time"
 
-	"github.com/alibaba/openyurt/pkg/projectinfo"
-
-	"github.com/alibaba/openyurt/pkg/yurttunnel/constants"
 	certificates "k8s.io/api/certificates/v1beta1"
 	"k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
@@ -35,6 +32,9 @@ import (
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog"
+
+	"github.com/alibaba/openyurt/pkg/projectinfo"
+	"github.com/alibaba/openyurt/pkg/yurttunnel/constants"
 )
 
 // YurttunnelCSRApprover is the controller that auto approve all
@@ -86,7 +86,7 @@ func (yca *YurttunnelCSRApprover) processNextItem() bool {
 	csr, err := yca.csrInformer.Lister().Get(csrName)
 	if err != nil {
 		runtime.HandleError(err)
-		enqueueObj(yca.workqueue, csr)
+		yca.workqueue.AddRateLimited(key)
 	}
 
 	if err := approveYurttunnelCSR(csr, yca.csrClient); err != nil {

